tools/driver/driver.cpp

```diff
@@ -110,63 +112,8 @@ static bool shouldRunAsSubcommand(StringRef ExecName,
 
 extern int apinotes_main(ArrayRef<const char *> Args);
 
-int main(int argc_, const char **argv_) {
-  PROGRAM_START(argc_, argv_);
-
-  SmallVector<const char *, 256> argv;
-  llvm::SpecificBumpPtrAllocator<char> ArgAllocator;
-  std::error_code EC = llvm::sys::Process::GetArgumentVector(
-      argv, llvm::ArrayRef<const char *>(argv_, argc_), ArgAllocator);
-  if (EC) {
-    llvm::errs() << "error: couldn't get arguments: " << EC.message() << '\n';
-    return 1;
-  }
-
-  // Expand any response files in the command line argument vector - arguments
-  // may be passed through response files in the event of command line length
-  // restrictions.
-  llvm::BumpPtrAllocator Allocator;
-  llvm::StringSaver Saver(Allocator);
-  llvm::cl::ExpandResponseFiles(
-      Saver,
-      llvm::Triple(llvm::sys::getProcessTriple()).isOSWindows() ?
-      llvm::cl::TokenizeWindowsCommandLine :
-      llvm::cl::TokenizeGNUCommandLine,
-      argv);
-
-  // Check if this invocation should execute a subcommand.
-  StringRef ExecName = llvm::sys::path::stem(argv[0]);
-  SmallString<256> SubcommandName;
-  if (shouldRunAsSubcommand(ExecName, SubcommandName, argv)) {
-    // We are running as a subcommand, try to find the subcommand adjacent to
-    // the executable we are running as.
-    SmallString<256> SubcommandPath(
-      llvm::sys::path::parent_path(getExecutablePath(argv[0])));
-    llvm::sys::path::append(SubcommandPath, SubcommandName);
-
-    // If we didn't find the tool there, let the OS search for it.
-    if (!llvm::sys::fs::exists(SubcommandPath)) {
-      // Search for the program and use the path if found. If there was an
-      // error, ignore it and just let the exec fail.
-      auto result = llvm::sys::findProgramByName(SubcommandName);
-      if (!result.getError())
-        SubcommandPath = *result;
-    }
-
-    // Rewrite the program argument.
-    argv[0] = SubcommandPath.c_str();
-    
-    // Execute the subcommand.
-    argv.push_back(nullptr);
-    ExecuteInPlace(SubcommandPath.c_str(), argv.data());
-
-    // If we reach here then an error occurred (typically a missing path).
-    std::string ErrorString = llvm::sys::StrError();
-    llvm::errs() << "error: unable to invoke subcommand: " << argv[0]
-                 << " (" << ErrorString << ")\n";
-    return 2;
-  }
-
+static int run_driver(StringRef ExecName,
+                       const SmallVectorImpl<const char *> &argv) {
```

Comment:
Nitpick: ArrayRef is a handy wrapper type for any array-like structure; please use it here and in `shouldRunAsSubcommand`.